### ISSUE TITLE
testing LINUXPBARelease.img.gz yields method status code INVALID_PARAMETER

### ISSUE THREAD

**Issue by brianjmurrell** [Drive-Trust-Alliance/sedutil#73]
So I have a machine with an OPAL drive in it:

```
# msed --scan

Scanning for Opal compliant disks
/dev/sda  2  INTEL SSDSC2BF480A5L                     LUDi    
No more disks present ending scan
```

(yeah, I know, `msed` is old(er) but it's what's on Fedora 23 at the moment).

But when I try to [test](https://github.com/Drive-Trust-Alliance/sedutil/wiki/Test-the-PBA) the [PBA](https://github.com/Drive-Trust-Alliance/exec/blob/master/LINUXPBARelease.img.gz?raw=true) I ender some random value for the pass-phrase and get:

```
ERR : method status code INVALID_PARAMETER
ERR : Session start Failed
ERR : Unlock failed - unable to set LockingRange 0 RW
```

And then it reboots only to repeat.

That said, I have a biospba-0.23beta.img around from when I did this the last time.  Is that the `SYSLINUX` based PBA?  Is there any reason I would not just use that instead since it seems faster than this LINUXPBARelease.img and doesn't require a reboot?

Is there any reason why the `SYSLINUX` PBA is not what is being used by this project by default?

Regardless, of course I am hesitant to enable encryption on this drive when the documented test procedure fails out of the gate.

Any ideas?

**Comment by brianjmurrell** [comment]
So, looking around a bit more I found the [FAQs](https://github.com/Drive-Trust-Alliance/sedutil/wiki/FAQ) and one of them describes:

INVALID_PARAMETER: The command sent to the drive is incorrectly formatted. This can be either a program or user error. The most likely error is that you are trying to issue a command to the ADMIN SP before it has been activated. Make sure you have run initialsetup before trying to setup the locking ranges.

But the [test procedure](https://github.com/Drive-Trust-Alliance/sedutil/wiki/Test-the-PBA) does not mention running `... -–initialsetup <password> <drive>` until after the place in the document where it suggests testing the PBA first.

Do I have to run `... -–initialsetup <password> <drive>` before I can test the PBA?

**Comment by brianjmurrell** [comment]
I hate to sound impatient but any ideas on this?  I have a machine in limbo (i.e. I don't want to do anything on it, encryption-wise or otherwise) until I can be sure the PBA is going to work for me.

I don't really want to go ahead and try to encrypt the drive without knowing I can decrypt it and I don't want to use the machine for anything (i.e. adding data that I don't have anywhere else) else in case some kind of error/failure in the encrypting part of it renders the contents un-decryptable.  Right now that machine has the content of the disk it's replacing so any fubars are recoverable although with more than just an insignificant time and effort.

**Comment by brianjmurrell** [comment]
So I took a chance... Running:

```
sedutil-cli -–initialSetup <password> <drive>
```

does not change the output of the PBA test.  It still produces the same errors.  If I try to run that again I get:

```
ERR : method status code NOT_AUTHORIZED
ERR : session start failed
ERR : One or more header fields have 0 length
ERR : EndSession Failed
ERR : takeOwnership failed
ERR : Initial setup failed - unable to take ownership
```

I'm guessing that operation is not idempotent.  Perhaps not at least until I can successfully authenticate to the drive, which the PBA test does not seem to be doing.

FWIW, with the PBA test, if I enter the password I previously used with ```--initialSetup```, I get the NOT_AUTHORIZED errors.  I am most definitely using the same password I used with ```--initialSetup```.  Then the machine reboots and I get the DTA PBA passphrase prompt again.

I'm 90% sure I didn't get this NOT_AUTHORIZED the first time I tried to test the PBA after running ```--initialSetup``` but just get getting the PBA passphrase prompt.

**Comment by brianjmurrell** [comment]
FWIW, it would be nice if the PBA stopped after emitting the errors and allowed one time to read them before rebooting the machine.  That's probably a separate ticket though.

**Comment by brianjmurrell** [comment]
I feel like I am getting further down a rathole here.  So here's some more information about the current state:

```
# msed --scan

Scanning for Opal compliant disks
/dev/sda  2  INTEL SSDSC2BF480A5L                     LUDi    
No more disks present ending scan
# msed --query /dev/sda

/dev/sda ATA INTEL SSDSC2BF480A5L                     LUDi     CVTRXXXXXXN480EGN  
TPer function (0x0001)
    ACKNAK = N, ASYNC = N. BufferManagement = N, comIDManagement  = N, Streaming = Y, SYNC = Y
Locking function (0x0002)
    Locked = N, LockingEnabled = Y, LockingSupported = Y, MBRDone = Y, MBREnabled = Y, MediaEncrypt = Y
Geometry function (0x0003)
    Align = Y, Alignment Granularity = 16 (8192), Logical Block size = 512, Lowest Aligned LBA = 0
SingleUser function (0x0201)
    ALL = N, ANY = N, Policy = Y, Locking Objects = 5
DataStore function (0x0202)
    Max Tables = 9, Max Size Tables = 10485760, Table size alignment = 1
OPAL 2.0 function (0x0203)
    Base comID = 0x0800, Initial PIN = 0x0, Reverted PIN = 0x0, comIDs = 1
    Locking Admins = 4, Locking Users = 9, Range Crossing = N

TPer Properties: 
  MaxMethods = 1  MaxSubpackets = 1
  MaxPacketSize = 8684  MaxPackets = 1  MaxComPacketSize = 8704
  MaxResponseComPacketSize = 8704  MaxSessions = 1  MaxReadSessions = 1
  MaxIndTokenSize = 8648  MaxAggTokenSize = 8648  MaxAuthentications = 3
  MaxTransactionLimit = 1  DefSessionTimeout = 900000  MaxSessionTimeout = 86400000
  MinSessionTimeout = 30000  DefTransTimeout = 1000  MaxTransTimeout = 1000
  MinTransTimeout = 100  MaxComIDTime = 60000  ContinuedTokens = 0
  SequenceNumbers = 1  AckNak = 0  Asynchronous = 0

Host Properties: 
  MaxMethods = 1  MaxSubpackets = 1
  MaxPacketSize = 2028  MaxPackets = 1  MaxComPacketSize = 2048
  MaxIndTokenSize = 1992  MaxAggTokenSize = 1992  SequenceNumbers = 1
```

Given that I have run ```--initialsetup <password> /dev/sda" but that the PBA keeps telling me NOT_AUTHORIZED, is there any way I can test if the password that I think should be the password is the password, without using the PBA?  Like an ```msed``` command that will confirm if the password is as I think it is?

**Comment by brianjmurrell** [comment]
Ahhh.

I did try:

```
# msed --setSIDPwd foobar <my_password> /dev/sda
```

to see what ```msed``` does if i use the wrong password and it reported:

```
- 23:39:32.811 ERR : method status code NOT_AUTHORIZED
- 23:39:32.811 ERR : Session start failed
- 23:39:32.844 ERR : One or more header fields have 0 length
- 23:39:32.844 ERR : EndSession Failed
```

Which is good.  So  I then tried:

```
# msed --setSIDPwd <my_password> <my_password> /dev/sda
```

and got no error so that would seem to indicate to me that ```<my_password>``` is the correct one.  Yet this is the same password that PBA test is reporting NOT_AUTHORIZED.

**Comment by brianjmurrell** [comment]
OK.  We can stand-down the red-alert on this one.

I bit the bullet, did a PSID revert on the drive and tested the PBA by actually installing it as [the process](https://github.com/Drive-Trust-Alliance/sedutil/wiki/Encrypting-your-drive) describes.  It worked.

I ended up having to configure the drive again and re-copy the data to the drive after encrypting it but it seems to be all working now.

I'd still like to understand the status of the syslinux version of the PBA vs. this Linux based one.  I liked that the syslinux one didn't reboot the machine after entering the password.

**Comment by r0m30** [comment]
Sorry for your issues.  I also liked that the syslinux PBA didn't reboot the system, I wrote it first and it worked fine on my machine, as sedutil was installed on more machines it became apparent that the SATA interface was not very robust.  It has a primitive SATA driver that works for some machines and doesn't for others.  I have tried to remote debug issues with it before with little success.  So at this point it is available to try and if it works great, if not then you will need to use the Linux based PBA.

We hope to develop a UEFI PBA similar to the syslinux PBA but work on that has not even started.

**Comment by brianjmurrell** [comment]
> it is available to try

Is this it [here](https://github.com/Drive-Trust-Alliance/exec/blob/master/biospba.img.gz?raw=true)?

Pressumably one can just load it over the LINUXPBA using ```sedutil-cli –-loadPBAimage <password> <pbafilename>  <drive>``` but what does one do if it fails to work?  Does one boot on the [Rescue Image](https://github.com/Drive-Trust-Alliance/exec/blob/master/Rescue.img.gz?raw=true) and then reload the LINUXPBA again?

Can that be done simply with ```sedutil-cli –-loadPBAimage <password> <pbafilename>  <drive>``` again or some kind of unlocking need to be done before that?

**Comment by r0m30** [comment]
Yes. that's it.

If it failed I'd boot the rescue image, unlock the drive and then reload the othe PBA.  I've never tried to write the PBA to a locked drive, so I can't say if it will work or not.

**Comment by dasois** [comment]
Same issue here.
Testet 1.12 BIOS, 1.12 UEFI64 and 1.10 biosbpa, all produce:
ERR : method status code NOT_AUTHORIZED
ERR : session start failed
ERR : One or more header fields have 0 length
ERR : EndSession Failed
with a reboot

This is a fresh Samsung 850 PRO SSD, that I did not encrypt previously. In a Thinkpad T550 Laptop.

It does not feel right to ignore and start with the encryption process, what shall I do?

**Comment by r0m30** [comment]
I don't think 1.10 biospba produced the same output as the other two.  The NOT_AUTHORIZED error means that the password you entered is not the correct password for the Admin1 authority.  Have you run intialsetup command?

**Comment by r0m30** [comment]
You don't need to run intiialsetup, you really shouldn't run it until AFTER you verify the PBA.  Initialsetup sets the MBR shadow on so the drive will not be accessible after a powercycle.  I was asking because the messages you get will be different if you have run initialsetup.

**Comment by Lutwick** [comment]
Hi, I also just tried the ["Test the PBA" described in the wiki](https://github.com/Drive-Trust-Alliance/sedutil/wiki/Test-the-PBA) and got the INVALID_PARAMETER error:

![bootup](https://cloud.githubusercontent.com/assets/4678285/18027462/29f47ad6-6c64-11e6-9692-92b8b6418180.png)
![error](https://cloud.githubusercontent.com/assets/4678285/18027466/300a3172-6c64-11e6-90bb-312d4287f30b.png)


C:\sedutil_WIN\Win32\Release>sedutil-cli.exe --scan

Scanning for Opal compliant disks
\\.\PhysicalDrive0 12  Samsung SSD 840 EVO 120GB                EXT0DB6Q
No more disks present ending scan



C:\sedutil_WIN\Win32\Release>sedutil-cli.exe --query \\.\PhysicalDrive0

\\.\PhysicalDrive0 ATA Samsung SSD 840 EVO 120GB                EXT0DB6Q S1D5NSAF519613P
TPer function (0x0001)
    ACKNAK = N, ASYNC = N. BufferManagement = N, comIDManagement  = N, Streaming = Y, SYNC = Y
Locking function (0x0002)
    Locked = N, LockingEnabled = N, LockingSupported = Y, MBRDone = N, MBREnabled = N, MediaEncrypt = Y
Geometry function (0x0003)
    Align = N, Alignment Granularity = 1 (512), Logical Block size = 512, Lowest Aligned LBA = 0
Opal V1.0 function (0x0200)
Base comID = 0x1004, comIDs = 1
DataStore function (0x0202)
    Max Tables = 9, Max Size Tables = 10485760, Table size alignment = 1
OPAL 2.0 function (0x0203)
    Base comID = 0x1004, Initial PIN = 0x0 , Reverted PIN = 0x0 , comIDs = 1
    Locking Admins = 4, Locking Users = 9, Range Crossing = N

TPer Properties:
  MaxComPacketSize = 66048  MaxResponseComPacketSize = 66048
  MaxPacketSize = 66028  MaxIndTokenSize = 65992  MaxPackets = 1
  MaxSubpackets = 1  MaxMethods = 1  MaxAuthentications = 5
  MaxSessions = 1  MaxTransactionLimit = 1  DefSessionTimeout = 0

Host Properties:
  MaxComPacketSize = 2048  MaxResponseComPacketSize = 2048
  MaxPacketSize = 2028  MaxIndTokenSize = 1992  MaxPackets = 1
  MaxSubpackets = 1  MaxMethods = 1


After reading this, I guess I should also just try to encrypt my disk.
But then the  ["Test the PBA"](https://github.com/Drive-Trust-Alliance/sedutil/wiki/Test-the-PBA) is really useless...?

**Comment by r0m30** [comment]
If you still have the time to update the doc on testing the PBA that would be great.

**Comment by r0m30** [comment]
PBA and doc updated
